### ISSUE TITLE
Too many request (HTTP 429 response)

### ISSUE THREAD

**Issue by palepupet** [f3ath/appnexusclient#11]
hi guys, 

Will there be a feature that will take into account the HTTP 429 response when too much query / minutes is done ?

thx

**Comment by f3ath** [comment]
Hi @palepupet, thanks for opening the issue. I'd be happy to address it if I had a test/dev account at appnexus to test the changes. Otherwise it will be shooting in the dark. Unfortunately I no longer have such an account. Is there any way you could assist me getting a developer's account?

**Comment by palepupet** [comment]
Hello,

Thank you for your answer. Unfortunately I can't give you an account
because the one I use belongs to the company I work for. However if you are
ok I could give you some help or documentation on how some aspect of the
API works.

​Let me know.



2018-05-30 20:22 GMT+02:00 Alexey <notifications@github.com>:

**Comment by f3ath** [comment]
Sure. Documentation may help as long as you can do beta testing for me :)

**Comment by palepupet** [comment]
Ok, let's do that.

Next time you want some doc, ask me.


2018-05-30 20:36 GMT+02:00 Alexey <notifications@github.com>:

**Comment by f3ath** [comment]
Alright. So as far as I understand, you're getting HTTP 429. Does the documentation say there is actually a limit of this kind? If so, what's the limit and how many RPS you're making?

**Comment by palepupet** [comment]
Yes,

There is a limit, 60 write / minute and 100 read / minute. Sometimes I need
to make thousands of calls. However it is possible, when we have an 429
http answer, to look in the header. there is a "retry-after" field and a
number. It is this number which allows to wait the necessary time before
being able to make a new call.

2018-05-30 20:55 GMT+02:00 Alexey <notifications@github.com>:

**Comment by palepupet** [comment]
If you want you can contact me on my email address profile. It would be more simple.

**Comment by f3ath** [comment]
Let's keep the communication here. Email is nice, but it would be hard to find in case we need it in the future.

So, looks like we're talking about adding a request queue to this library. I would rather not do that to keep the library small and concerns separated. This is a simple synchronous library, so adding support for async or queued calls would make it unnecessary complex. Instead, what I can do here is to throw a special kind of domain exception, say `RateLimitExceededException` in case the server responds with 429. So then you can do something like this

```php

$job = function(...) use (...) {
  $response = $appnexus->call(...);
  // process $response according to your application logic
};
try {
  $job();
} catch (RateLimitExceededException $e) {
  addJobToQueue($job, $e->getRetryAfter());
}
```

So that your business logic can decide how to queue and process the requests, while the library will keep the implementation details (like where to get the retry-after value from) hidden.

**Comment by palepupet** [comment]
Yes, I think catch the 429 http response is a better idea than the other too. So acting like that, if there is a RateLimitExceededException catched, the request is place to a queue and after x seconds, found in the retry field, the same request bloqued is resend. This is it ?

**Comment by f3ath** [comment]
Yes. It means the library will just throw the exception, while the queuing/rescheduling logic will be somewhere in your application. If you're OK with this, I need to see the corresponding part of the documentation. Can you provide me with the docs please?

**Comment by palepupet** [comment]
No problems, there it is :
https://wiki.appnexus.com/display/api/API+Best+Practices#APIBestPractices-Throttleyourcalls

**Comment by f3ath** [comment]
@palepupet which version of php are you using?

**Comment by palepupet** [comment]
I'm using php 5.6.31

**Comment by palepupet** [comment]
Hi, 
I'm coming back to you to see if everything is going well and if you need more informations. I also wanted to ask you if it was possible to do, as for call limits, to make a special exception for server 500 errors.
let me know.
thx
